gluejobutils/scd.py

```diff
@@ -0,0 +1,209 @@
+    dea_record_end_datetime is set to the next records dea_record_start_datetime. If the dea_record_end_datetime is last row of that partition then it is set to 
+    the default static_record_end_datetime (i.e. 2999-01-01 00:00:00)
+    """
+    start_col = 'dea_record_start_datetime' 
+    end_col = 'dea_record_end_datetime'
+    
+    df = df.withColumn(end_col, F.expr("lead({}) OVER (PARTITION BY {} ORDER BY {})".format(start_col, partition_by, order_by)))
+    
+    # Set nulls to default record end date value
+    df = df.withColumn(end_col, F.expr("CASE WHEN {0} IS NULL THEN to_timestamp('{1}', '{2}') ELSE {0} END".format(end_col, static_record_end_datetime, standard_datetime_format)))
+    
+    # Filter out redundent records
+    df = df.filter("{} <> {}".format(start_col, end_col))
+
+    return df
+
+def upsert_table_by_record(spark, new_df, table_db_base_path, update_by_cols, partition_path = '', coalesce_size = 4, update_latest_records_only = True, write_back_table_db_path = False) :
+    """
+    Updates the data that current exists at table_db_base_path with new_df.
+
+    new_df should have the exact same columns and meta as table_db_base_path (therefore record start and end dates should have been
+    applied to new_df.
+
+    table at table_db_base_path is read in and joined to new_df on update_by_cols. Any rows from table_db_base_path 
+    that do not match to new_df are written to a partition called dea_record_update_type (set to old). These are records that are unchanged. 
+
+    The rest of table_db_base_path is appended to new_df as these rows also exist in new_df so we need to recalculate SCD2.
+
+    The SCD2 (specifically dea_record_start_date and dea_record_end_date) are recalculated for the appended df. SCD2 are recalculated
+    using the update_by_cols variable. This data is then written the same tmp directory but with the partition dea_record_update_type set to new.
+
+    If no file exists in table_db_base_path then this function writes new_df straight to the tmp folder with partition dea_record_update_type set to new.
+
+    update_latest_records_only means that only latest records are called from current table. Set to false if you wish to back fill old update.
+    """
+    if not isinstance(update_by_cols, list) :
+        raise TypeError('update_by_cols must be a list of column names')
+
+    bucket, key = s3_path_to_bucket_key(table_db_base_path)
+    table_base_path, table_name = os.path.split(remove_slash(key))
+
+    table_db_path = add_slash(os.path.join(table_db_base_path, partition_path))
+    tmp_table_db_path = os.path.join(table_base_path, table_name + '_tmp', partition_path)
+    tmp_table_db_path_old_partition = add_slash(os.path.join('s3://', bucket, tmp_table_db_path, 'dea_record_update_type=old'))
+    tmp_table_db_path_new_partition = add_slash(os.path.join('s3://', bucket, tmp_table_db_path, 'dea_record_update_type=new'))
+
+    if folder_contains_only_files_with_extension(table_db_path) :
```

Comment:
I'm not sure I understand why the `if` is conditional on `folder_contains_only_files_with_extension(table_db_path) `

Could we split into two i.e
```
variable_with_meaningulname_which_corresponds_to_test = folder_contains_only_files_with_extension(table_db_path) 

if variable_which_corresponds_to_test:

```